lib/tests/formulae.rb

```diff
@@ -530,18 +530,10 @@ def formula!(formula_name, args:)
           bottle_reinstall_formula(formula, new_formula, args:)
         end
         @built_formulae << formula.full_name
-        test("brew", "linkage", "--test", named_args: formula_name, ignore_failures:, report_analytics: true)
+        test("brew", "linkage", "--test", "--strict", named_args: formula_name, ignore_failures:, report_analytics: true)
```

Comment:
This will break `gcc` and `julia` builds. Haven't really been able to work out how to prevent their failures under `--strict` (because they have files with missing `rpath`s). The warning is, AFAICT, harmless for `julia`, but not necessarily for GCC (though it happens for a compatibility library that is either deprecated or about to be).

CC @fxcoudert for thoughts.

Reply:
For compiler runtime libraries, the lack of rpaths is not necessarily an issue. For GCC, the compatibility library is not going away, but the lack of rpath (a historical feature) is not a problem since it is only supposed to be linked by gcc itself, which will add the correct path in the executable.

I think the solution is to differentiate between different audits:
- lack of rpath in some cases is not a fatal flaw, and should not be an error
- linking to a library not in a dependency should be an error (in strict mode)

Reply:
> For GCC, the compatibility library is not going away, but the lack of rpath (a historical feature) is not a problem since it is only supposed to be linked by gcc itself, which will add the correct path in the executable.

Sadly it is a problem for us, because GCC adds `LC_RPATH` commands referencing `Cellar` paths to the binaries it links, so we need the compatibility library to have a correct (relative) `LC_RPATH` command.

Reply:
CC @Bo98 too.

What is the proposed solution here? I see a few:
1) don't do this
2) do this and have an allowlist for `gcc`/`julia` in Homebrew/brew
3) do this and have an allowlist for `gcc`/`julia` in Homebrew/homebrew-test-bot
4) do this but with something else too

Reply:
Almost feels like we maybe need a middle ground:

* `--test` that `brew install` runs`
* `--test` that `brew test-bot` run

Reply:
> Either the `gcc`/`julia` is wrong or the rpath check is wrong.

I agree with this. I'm not sure what the point is in having warnings like this that just get ignored. What purpose do these checks serve and why do GCC/Julia not need/comply with them?

> Almost feels like we maybe need a middle ground:

Yeh, I'm leaning in this direction too: a new flag in Homebrew/brew that's called here?

Reply:
> I agree with this. I'm not sure what the point is in having warnings like this that just get ignored. What purpose do these checks serve and why do GCC/Julia not need/comply with them?

GCC should comply with them. Arguably Julia should as well, but that's less of a problem. But even if we assume they don't need to comply -- these checks are still useful for dozens of other formulae I've fixed where they are problems. (For examples, `grep` for `CMAKE_INSTALL_RPATH` -- many of them are caught by `test` blocks, but not all of them.)

> Yeh, I'm leaning in this direction too: a new flag in Homebrew/brew that's called here?

`--strict` and `--very-strict`? `--moderate` and `--strict`?

Reply:
> if we assume they don't need to comply -- these checks are still useful for dozens of other formulae I've fixed where they are problems

If they don't: this seems very close to being a candidate for moving to `brew audit` where we do have allowlists and selective `--only/--except` logic.

Would also simplify `brew linkage` to not need three modes.

Reply:
Responding to:

> why do GCC/Julia not need/comply with them?

and 

> GCC should comply with them

GCC ships a fully functional compiler, the lack of rpath in libgcc_s is design choice and not a bug, and it does not trigger any issue. libgcc_s only exists to be linked by GCC itself, and GCC will provide the necessary information when it links.

The audit is useful for a lot of standalone libraries, which is maybe 99% of Homebrew, but libgcc_s is not a standalone library. Neither are the julia shared objects.

So for me the root cause is: the audit assumes that every shared object / dylib is standalone shared library. That is true for most software, but not always the case.

Reply:
> GCC will provide the necessary information when it links.

This is not correct. GCC encodes `LC_RPATH`s that reference `Cellar` paths. These will break as soon as we rev-bump or upgrade GCC:

```
❯ gcc-14 -xc - <<<'int main(){}'; otool -l a.out | rg -A2 LC_RPATH
          cmd LC_RPATH
      cmdsize 32
         path @loader_path (offset 12)
--
          cmd LC_RPATH
      cmdsize 96
         path /opt/homebrew/Cellar/gcc/14.1.0/lib/gcc/current/gcc/aarch64-apple-darwin23/14 (offset 12)
--
          cmd LC_RPATH
      cmdsize 64
         path /opt/homebrew/Cellar/gcc/14.1.0/lib/gcc/current/gcc (offset 12)
--
          cmd LC_RPATH
      cmdsize 64
         path /opt/homebrew/Cellar/gcc/14.1.0/lib/gcc/current (offset 12)
```

Reply:
@carlocab we've upgraded GCC from 13 to 14 recently, and none of the dependencies needed to be rebuilt (at least for that reason).

Reply:
That's because we don't have any formulae that link with `libgcc_s.1`. We used to, and it did cause problems for that formula, but you fixed it upstream: https://github.com/JuliaLang/julia/pull/46240

The fact is that if you link with Homebrew's `libgcc_s.1`, `pkg_version` changes to GCC will break you.

Reply:
Nothing is supposed to link with `libgcc_s.1` anyway. We provide it (only on Intel) for code that was already compiled against older GCC (much older).

Anyway, if you fix `libgcc_s.1` rpath in GCC (assuming that doesn't break executables linked against earlier libraries?) the same is true of Julia. It's not a bug, it's just a different use case for shared objects. You can also have dylibs that are dynamically loadable modules, but need no rpath and could potentially have none.

Reply:
> Nothing is supposed to link with `libgcc_s.1` anyway. We provide it (only on Intel) for code that was already compiled against older GCC (much older).

Yes. But the fact still remains that our `libgcc_s.1` is broken -- we should either fix it or get rid of it.

> Anyway, if you fix `libgcc_s.1` rpath in GCC (assuming that doesn't break executables linked against earlier libraries?)

Given how `dyld` handles `LC_RPATH` commands, I really don't see how it could break the aforementioned executables.

> the same is true of Julia. It's not a bug, it's just a different use case for shared objects. You can also have dylibs that are dynamically loadable modules, but need no rpath and could potentially have none.

On macOS, if a library is intended to be used only as a dynamically loadable module, then it should be built with `-bundle` and not `-dynamiclib`/`-shared`. (If these were built as `-bundle`s then the linkage failure will not happen.) To me, failing to do this is a bug.

Reply:
> If they don't: this seems very close to being a candidate for moving to `brew audit` where we do have allowlists (`audit_exceptions`) and selective `--only/--except` logic.
> 
> Would also simplify `brew linkage` to not need three modes.

I'd be 👍🏻 on this.

Simpler still, though, may be just adding the allowlist in the Homebrew/brew code itself. If we don't want/expect this list to grow: this seems fine and easier than moving this to a post-install audit.

Reply:
GCC warning is not really fixed -- it just goes away on Linux, but will still happen on macOS.

Reply:
Ok, thanks. Closing because it's not clear what needs to be done here.

Reply:
Here are our options to unblock this change:

1. Fix GCC and Julia so that they don't fail `brew linkage --test --strict`.
2. Create an allowlist for GCC and Julia that allows us to ignore `brew linkage --test --strict` failures.
3. Add a new flag to `brew linkage` that enables making indirect linkage an error in `test-bot`.
4. Get rid of the missing `RPATH` check in `brew linkage`.

(1) can be done easily for GCC with either a slightly hacky solution (adding RPATHs), or with `rm lib/"libgcc_s.1.dylib"`. It is likely to be fragile for Julia.

(2) is subject to @Bo98's criticism [above](https://github.com/Homebrew/homebrew-test-bot/pull/1059#discussion_r1599912129):

> However, more generally: I don't consider `brew linkage` false positives acceptable even with an allowlist. Either the `gcc`/`julia` is wrong or the rpath check is wrong. Past attempts to add allowlists to `brew linkage` have all ended up being removed for being wrong (see `ignore_missing_libraries` stuff).

(For clarity: I think `gcc`/`julia` is wrong, and have explained why above, but I believe FX still disagrees with me.)

(3) could make the flags for `brew linkage` confusing/complicated, if we went down a `--strict` and `--stricter` route for the flags. It might be more viable if we did `--strict`, which enables all strict checks, and `--strict=foo,bar`, where `foo` and `bar` are the strict checks we want to run (e.g. `--strict=indirect-linkage`).

(4) would be easy, but it limits our ability to proactively fix issues that could be a problem for users.

Reply:
- For GCC, removing the compat library is wrong, but fixing its rpath is acceptable. I don't think it's a bug and the audit is being overly picky, but it's not a big deal.
- Re. Julia, has an issue been opened upstream?

Reply:
Why isn't the fact that user binaries that link to `libgcc_s.1` will break when the GCC `pkg_version` changes a bug?

I'm also sceptical that the lack of an `LC_RPATH` command in `libgcc_s.1` is deliberate. From the libraries in GCC, here are the ones with an `@rpath` reference:

```
❯ fd --type=file --extension=dylib -x sh -c 'otool -L {} | rg -q rpath && echo {}' | xargs otool -L
./libgcc_s.1.dylib:
        /usr/local/opt/gcc/lib/gcc/current/libgcc_s.1.dylib (compatibility version 1.0.0, current version 1.1.0)
        @rpath/libgcc_s.1.1.dylib (compatibility version 1.0.0, current version 1.1.0, reexport)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.100.3)
./libobjc-gnu.4.dylib:
        /usr/local/opt/gcc/lib/gcc/current/libobjc-gnu.4.dylib (compatibility version 5.0.0, current version 5.0.0)
        @rpath/libgcc_s.1.1.dylib (compatibility version 1.0.0, current version 1.1.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.100.3)
./libgfortran.5.dylib:
        /usr/local/opt/gcc/lib/gcc/current/libgfortran.5.dylib (compatibility version 6.0.0, current version 6.0.0)
        @rpath/libquadmath.0.dylib (compatibility version 1.0.0, current version 1.0.0)
        @rpath/libgcc_s.1.1.dylib (compatibility version 1.0.0, current version 1.1.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.100.3)
./libstdc++.6.dylib:
        /usr/local/opt/gcc/lib/gcc/current/libstdc++.6.dylib (compatibility version 7.0.0, current version 7.32.0)
        /usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
        @rpath/libgcc_s.1.1.dylib (compatibility version 1.0.0, current version 1.1.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.100.3)
```

Of these, we can see which ones have `LC_RPATH` commands:
```
❯ fd --type=file --extension=dylib -x sh -c 'otool -L {} | rg -q rpath && echo {}' | while read dylib; do echo "===== $dylib ====="; otool -l $dylib | rg -A2 LC_RPATH || echo none; echo; done
===== ./libgcc_s.1.dylib =====
none

===== ./libobjc-gnu.4.dylib =====
          cmd LC_RPATH
      cmdsize 32
         path @loader_path (offset 12)

===== ./libgfortran.5.dylib =====
          cmd LC_RPATH
      cmdsize 32
         path @loader_path (offset 12)

===== ./libstdc++.6.dylib =====
          cmd LC_RPATH
      cmdsize 32
         path @loader_path (offset 12)
```

So, all libraries with an `@rpath` reference has the correct `LC_RPATH` command, _except for_ `libgcc_s.1`.

I suppose you could argue that `libgcc_s.1` is somehow special, but I don't see why it would be. In particular, I think the argument that

> the lack of rpath (a historical feature) is not a problem since it is only supposed to be linked by gcc itself, which will add the correct path in the executable

also applies to `libstdc++`, etc., and yet they all contain the `LC_RPATH` command anyway. What makes `libgcc_s.1` special?

Reply:
> * Fix GCC and Julia so that they don't fail `brew linkage --test --strict`.
> * Create an allowlist for GCC and Julia that allows us to ignore `brew linkage --test --strict` failures.
> * Get rid of the missing `RPATH` check in `brew linkage`.

Fine with any of these with a mild preference for 1st > 3rd > 2nd.

> * Add a new flag to `brew linkage` that enables making indirect linkage an error in `test-bot`.

Would rather not do this.